src/charm.py

```diff
@@ -0,0 +1,250 @@
+        vault_config = {
+            "backend": backends,
+            "listener": {"tcp": {"tls_disable": True, "address": f"[::]:{self.VAULT_PORT}"}},
+            "default_lease_ttl": self.model.config["default_lease_ttl"],
+            "max_lease_ttl": self.model.config["max_lease_ttl"],
+            "disable_mlock": True,
+            "cluster_addr": f"http://{self._bind_address}:{self.VAULT_CLUSTER_PORT}",
+            "api_addr": f"http://{self._bind_address}:{self.VAULT_PORT}",
+        }
+
+        return Layer(
+            {
+                "summary": "vault layer",
+                "description": "pebble config layer for vault",
+                "services": {
+                    "vault": {
+                        "override": "replace",
+                        "summary": "vault",
+                        "command": "/usr/local/bin/docker-entrypoint.sh server",
+                        "startup": "enabled",
+                        "environment": {
+                            "VAULT_LOCAL_CONFIG": json.dumps(vault_config),
+                            "VAULT_API_ADDR": f"http://[::]:{self.VAULT_PORT}",
+                        },
+                    }
+                },
+            }
+        )
+
+    def _new_policy_action(self, event) -> None:
+        """Handler triggerred whenever an operator uses the "new-policy" action.
+
+        Args:
+            event: Juju event
+
+        Returns:
+            None
+        """
+        self.vault.set_policy(
+            name=event.params["name"], hcl=event.params["hcl"], backend=event.params["backend"]
+        )
+
+    def _new_app_role_action(self, event):
+        """Handler triggerred whenever an operator uses the "new-app-role" action.
+
+        Args:
+            event: Juju event.
+
+        Returns:
+            None
+        """
+        self.vault.create_role(
+            approle_name=event.params["name"],
+            policy_name=event.params["policy"],
+            cidr=event.params["cidr"],
+        )
+
+    def _get_token_action(self, event):
+        """Handler triggerred whenever an operator uses the "get-token" action.
+
+        Args:
+            event: Juju event. Event parameter must contain "name" and "cidr".
+
+        Returns:
+            None
+        """
+        response = self.vault.get_token(name=event.params["name"], cidr=event.params["cidr"])
+        event.set_results({"token": response["wrap_info"]["token"]})
+
+    def _get_root_token_action(self, event):
+        """Handler triggerred whenever an operator uses the "get-root-token" action.
+
+        The token can only be retrieved once. All the times after, "NA" will be returned to the
+        operator.
+
+        Args:
+            event: Juju event.
+
+        Returns:
+            None
+        """
+        if not self._stored.root_token_provided:
+            event.set_results({"token": self._stored.root_token})
+            self._stored.root_token_provided = True
```

Comment:
This works fine, but it might be safer to just set the `self.stored.root_token = ""` and conditionally return based on it being truthy? That way if the `StoredState` is dropped, we're less likely to try and provide the token if `root_token_provided != True`?

Reply:
The root token is used by the charm when making api calls to Vault (ex. for creating certificates). Here if we follow your suggestion, if the operator retrieves the root token, charms won't be able to get certificates from Vault.

Reply:
It's entirely possible for the charm to generate it's own token that can be used to perform it's tasks (creating certs), after which it would be fine to wipe the root token from the charm. That said, at the point where the vault unit can do everything it needs, it has a fairly powerful token anyway

Reply:
@ChrisMacNaughton Would you recommend creating such a token or sticking with using the root token?

Reply:
It would be more secure to have a dedicated token used for this that only has the required permissions; however, it could make charm-upgrades difficult if the needed permissions change.

To address @jnsgruk's comment regarding being able to retrieve the token if `StoredState` is dropped: the charm would forget the token too in that case, maybe we need an additional action to re-upload a root-token so that the charm can continue functioning if `StoredState ` is dropped.

Reply:
So following feedback from @javacruft , the charm nows very much mimics the behaviour of the charm-vault (machine charm). The operator has to manually initialize and unseal Vault unless the `totally-unsecure-auto-unlock` config is set to True. @jnsgruk , @ChrisMacNaughton @javacruft If you can re-review the PR now that these new changes are implemented please.